RFS/src/main/java/com/rfs/common/DocumentReindexer.java

```diff
@@ -1,47 +1,36 @@
 package com.rfs.common;
 
 import java.time.Duration;
-import java.util.Base64;
 import java.util.List;
 
-import io.netty.buffer.Unpooled;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.lucene.document.Document;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
-import reactor.netty.http.client.HttpClient;
 import reactor.util.retry.Retry;
 
 
 public class DocumentReindexer {
     private static final Logger logger = LogManager.getLogger(DocumentReindexer.class);
     private static final int MAX_BATCH_SIZE = 1000; // Arbitrarily chosen
 
-    public static void reindex(String indexName, Flux<Document> documentStream, ConnectionDetails targetConnection) throws Exception {
-        String targetUrl = "/" + indexName + "/_bulk";
-        HttpClient client = HttpClient.create()
-            .host(targetConnection.hostName)
-            .port(targetConnection.port)
-            .headers(h -> {
-                h.set("Content-Type", "application/json");
-                if (targetConnection.authType == ConnectionDetails.AuthType.BASIC) {
-                    String credentials = targetConnection.username + ":" + targetConnection.password;
-                    String encodedCredentials = Base64.getEncoder().encodeToString(credentials.getBytes());
-                    h.set("Authorization", "Basic " + encodedCredentials);
-                }
-            });
+    public static Mono<Void> reindex(String indexName, Flux<Document> documentStream, ConnectionDetails targetConnection) throws Exception {   
```

Comment:
Returning something for every future operation makes a ton of sense.  Take a look at my [PR to overhaul our async promises](https://github.com/opensearch-project/opensearch-migrations/pull/614).  Returning these so that they can be chained, even if just for monitoring is a good move.  
I'm not familiar with the reactor client, but there may be a longer term question about whether you want/need to embrace reactor primitives vs more universal ones (like CompletableFuture), which we have significant diagnostic support for at the moment.  Longer term (PR-614 switched out the netty future stuff for CompletableFutures, so it's a straightforward if borderline arduous task).

Reply:
I'm not sure myself, yet.  I think there is a limit to how far up the call stack this stuff will propagate; I've hit that limit in some local changes I have that aren't in this PR.  Definitely something to keep an eye on, agree!

Took a look at the referenced PR; don't think the RFS behavior is complex enough to need that level of abstraction.  I could be proven wrong, but it feels so far like we can get by with just using the Reactor primitives?

Reply:
Having a chain of 2-3 dependencies may be complex enough if you want to make sure that things aren't wedging forever or unintentionally holding up other calls.  I'm not sure where you're at, but if you have a web-client, you'll have more than that.  The reason to use something like TrackedFutures is that if you have > 1K activities per second, you're not really going to want to log enough details that if one goes wrong, you can find out what the issue is.  That would be slow and expensive (in CPU/storage + user time).  The approach that the replayer now takes is to keep those graphs around, then render only the ones that are becoming a problem.

It would be good to check the impact of TrackedFutures before putting it into more parts of the solution (it hasn't gone into the proxy yet).

This thread of comment came up because of the return type.  If you're worried about how high your reactor client is going, you could convert a mono object to a CompletableFuture (like what is done in NettyFutureBinders).

Reply:
I've actually found how high up the Rector types go at this point; it's not as high as I feared and actually makes sense (to me, at least).  You'll get a good look when I put out the next PR soon w/ the RfsWorker coded up to perform the Snapshot Phase.